### DISCUSSION TITLE
Add a link to a PDF File on a Specific Page while clipping text using tiddlyclip from localpdf file embedded in TW

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tiddlyclip-plugin#73] [Q&A] [answered]
I found these data related to [Link to a PDF File on a Specific Page](https://kookma.github.io/TW-Scripts/#Link%20to%20a%20PDF%20File%20on%20a%20Specific%20Page:%24%3A%2FAdvancedSearch%20%24%3A%2Fkookma%2Fmacros%2Fsearchox%20%5B%5BExample%2036%3A%20Global%20Macro%5D%5D%20exmp100%2Fmacro%2Fbookmarklet%20%5B%5BLink%20to%20a%20PDF%20File%20on%20a%20Specific%20Page%5D%5D%20demo%2Fexmp026%2Flink-pdf-use-tiddlerpath%20demo%2Fexmp026%2Flink-pdf-to-page%20%5B%5BSearch%20Wiki%5D%5D%20%5B%5BSearch%20in%20Tiddlywiki%20Toolmap%5D%5D%20%5B%5BSearch%20in%20%20Official%20Tiddlywiki%20Site%5D%5D%20%5B%5BHot%20Resources%5D%5D%20%5B%5BLatest%20Topics%5D%5D%20History%20Home%20%5B%5BTasks%20Explorer%5D%5D) in TW scripts by Kookma

1. [One](https://kookma.github.io/TW-Scripts/#demo%2Fexmp026%2Flink-pdf-to-page:%24%3A%2FAdvancedSearch%20%24%3A%2Fkookma%2Fmacros%2Fsearchox%20%5B%5BExample%2036%3A%20Global%20Macro%5D%5D%20exmp100%2Fmacro%2Fbookmarklet%20%5B%5BLink%20to%20a%20PDF%20File%20on%20a%20Specific%20Page%5D%5D%20demo%2Fexmp026%2Flink-pdf-use-tiddlerpath%20demo%2Fexmp026%2Flink-pdf-to-page%20%5B%5BSearch%20Wiki%5D%5D%20%5B%5BSearch%20in%20Tiddlywiki%20Toolmap%5D%5D%20%5B%5BSearch%20in%20%20Official%20Tiddlywiki%20Site%5D%5D%20%5B%5BHot%20Resources%5D%5D%20%5B%5BLatest%20Topics%5D%5D%20History%20Home%20%5B%5BTasks%20Explorer%5D%5D)
2. [Two](https://kookma.github.io/TW-Scripts/#demo%2Fexmp026%2Flink-pdf-use-tiddlerpath:%24%3A%2FAdvancedSearch%20%24%3A%2Fkookma%2Fmacros%2Fsearchox%20%5B%5BExample%2036%3A%20Global%20Macro%5D%5D%20exmp100%2Fmacro%2Fbookmarklet%20%5B%5BLink%20to%20a%20PDF%20File%20on%20a%20Specific%20Page%5D%5D%20demo%2Fexmp026%2Flink-pdf-use-tiddlerpath%20demo%2Fexmp026%2Flink-pdf-to-page%20%5B%5BSearch%20Wiki%5D%5D%20%5B%5BSearch%20in%20Tiddlywiki%20Toolmap%5D%5D%20%5B%5BSearch%20in%20%20Official%20Tiddlywiki%20Site%5D%5D%20%5B%5BHot%20Resources%5D%5D%20%5B%5BLatest%20Topics%5D%5D%20History%20Home%20%5B%5BTasks%20Explorer%5D%5D)

I was able to make the first pdf macro working in my local pdf file embedded using tiddlyclip localrules.

Then i created new rules

![image](https://user-images.githubusercontent.com/67494083/153269161-bcddca50-4868-4616-9d0f-3953d2fec50c.png)

![image](https://user-images.githubusercontent.com/67494083/153270170-40745da8-0b06-46e8-94ab-731862af09eb.png)

![image](https://user-images.githubusercontent.com/67494083/153269263-b5596eb5-442f-4cf7-9b5c-2f58dbb35e4c.png)

But i need to get the file path of the pdf file. Is it possible to extract the file path of the local pdf while clipping text from the pdf tiddler. I tried to use the `_canonical_uri` field of the pdf tiddler. But was unable to extract it. So i tried to manually copy paste the `_canonical_uri` field, but i found that the file path shown in the `_canonical_uri` field is slightly different from the file path copied from the browsers url field. The macro is working only if the file path copied from the browsers url field is used

**Comment by buggyj** [comment, marked as answer]
Are you clipping text from a pdf displayed in the browser and creating a link to that pdf/page?

**Reply by arunnbabu81** [reply]
> Are you clipping text from a pdf displayed in the browser and creating a link to that pdf/page?

No. I used the localpdf option of tiddlyclip and embedded the pdf in the wiki.

Clipping text from the pdf opened from the browser seems like a better idea than embedding the pdf in the wiki using localpdf rule?

**Comment by buggyj** [comment]
So you already have the pdf inside your tiddlywiki (using a canonical uri)and snipping text from there... and want to add this macro when clipping the text.

**Reply by arunnbabu81** [reply]
> So you already have the pdf inside your tiddlywiki (using a canonical uri)and snipping text from there... and want to add this macro when clipping the text.

Yes. But the clipping text from pdf displayed in the browser will be sufficient for me if that is an easier solution

**Reply by arunnbabu81** [reply]
> So you already have the pdf inside your tiddlywiki (using a canonical uri)and snipping text from there... and want to add this macro when clipping the text.

Is it possible if I want to clip from a pdf embedded within my wiki

**Comment by buggyj** [comment]
I will have to reply tomorrow

**Comment by arunnbabu81** [comment]
I tried to clip the text from the pdf displayed within the browser (not embedded within my TW). 
For this i made some rules.


![image](https://user-images.githubusercontent.com/67494083/153452809-1c05507e-66b5-4ea4-a85d-4547a6e56300.png)

![image](https://user-images.githubusercontent.com/67494083/153452933-3af87dfa-0118-40be-b280-fecfe1d13215.png)

![image](https://user-images.githubusercontent.com/67494083/153452986-95e1fe05-6187-40c0-934f-eaf74c76b406.png)

In this, the pop up window to input the userstring is not coming when I clip text.
I tested my other snip rules with userstring - these rules are not working with pdfs displayed within the browser, but these rules does work on other webpages.

**Reply by buggyj** [reply]
firefox does not allow pop up windows with pdf. You can put the page-number
in the note box in the tiddlyclip dropdown and use `((*@note*))`

**Reply by arunnbabu81** [reply]
> firefox does not allow pop up windows with pdf. You can put the page-number
> in the note box in the tiddlyclip dropdown and use `((*@note*))`

It’s working now…is there any way to change the position of the `link to the pdf page` created while clipping text

currently it’s positioned below the clipped  text

**Reply by arunnbabu81** [reply]
> firefox does not allow pop up windows with pdf. You can put the page-number in the note box in the tiddlyclip dropdown and use `((*@note*))`

Another problem is that text is not getting clipped whenever I try to clip after the first clip. 
For the first clip, both text and link to the pdf page comes.
For the next clips only the link to the pdf page is coming, no text is getting clipped. Also the links to pdf pages are getting arranged in the same line.

![image](https://user-images.githubusercontent.com/67494083/153468542-3778e924-e8fa-4d8f-a836-bffe10bc59e7.png)

**Reply by buggyj** [reply]
the code in `field initials` is only run one time when the tiddler is created.

you can move the `((*@text*))` from the `field initials` and using it in the `body` eg

`><<pdflink ............>>\n\n((*@text*))\n\n`

**Reply by arunnbabu81** [reply]
> the code in `field initials` is only run one time when the tiddler is created.
> 
> you can move the `((*@text*))` from the `field initials` and using it in the `body` eg
> 
> `><<pdflink ............>>\n\n((**@text))\n\n`

Now its working.
Is it possible to create a line break between the link and the clipped text.
Is it possible to bring the link to the left side of the clipped text?

**Reply by buggyj** [reply]
I cannot visualize what you want, create a mock up (in a tiddler) of what you want to see and put it here.

**Reply by arunnbabu81** [reply]
> I cannot visualize what you want, create a mock up (in a tiddler) of what you want to see and put it here.

![image](https://user-images.githubusercontent.com/67494083/153492897-67a86a9c-161a-4b37-9699-ddcad2b0120f.png)

In this image the first sentence of the clipped text is seen in the same line as the link. I want the clipped text to be separate from the link.

![image](https://user-images.githubusercontent.com/67494083/153493243-19ce0f2e-455b-4256-a041-b6301f5fbb72.png)

I want the link to be in the positioned at the level of the blue circle (to the left side of the clipped text) , not like the red circle

**Reply by buggyj** [reply]
what wiki text do you want to use to achieve this?

**Reply by arunnbabu81** [reply]
> what wiki text do you want to use to achieve this?

Can something like this 

```
@@float:right;
<span>
</span>
@@
```
be added to the pdf macro part in this code


`((*@text*))\n\n ><<pdflink label:"((*@note*))" path:"((*@pageRef*))" ((*@note*))>>`

so that the link floats to the right of the clipped text

**Reply by buggyj** [reply]
That won't work well.


you could make a table:

remove the contents of the 'body' column
create a new tiddler called `pdfFieldModifiers` with type `application/x-bclip`
with this content

`$text=|<<pdflink ............>>|((*@text*))|\n`

add `pdfFieldsModifiers` to the column `fieldmodifiers`

**Reply by arunnbabu81** [reply]
> That won't work well.
> 
> you could make a table:
> 
> remove the contents of the 'body' column create a new tiddler called `pdfFieldModifiers` with type `application/x-bclip` with this content
> 
> `$text=|<<pdflink ............>>|((*@text*))|\n`
> 
> add `pdfFieldsModifiers` to the column `fieldmodifiers`

This UI looks much better.

![image](https://user-images.githubusercontent.com/67494083/153620855-fc1978e3-c0fe-4cd2-bd6d-a9608a7084eb.png)

But the already clipped text gets gets replaced whenever i clip the next one.

These are the rules i use.

![image](https://user-images.githubusercontent.com/67494083/153621095-e5d4e11d-c87c-4c7b-89e1-36879fe1e85d.png)


![image](https://user-images.githubusercontent.com/67494083/153621194-8a4353a3-36cb-4a3e-b4f8-1893e0fd9830.png)


![image](https://user-images.githubusercontent.com/67494083/153621327-2c03e244-eb43-48f9-8f3d-26af8ce8294e.png)


![image](https://user-images.githubusercontent.com/67494083/153621402-af302d7e-6b20-47de-b716-554993ae13dd.png)

**Reply by buggyj** [reply]
the newline characters need removing from the text.
you will need the lastest version of tiddlyclip (2.0.01) 

add this line at the begining of pdfFieldModifiers
`@text=((*@removelf(@text)*))`
and change 
`$text=|><pdflink.....`
to 
`$text=((*$text*))|><pdflink........`

you will need this macro

[removelf.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/8051306/removelf.js.json.zip)

**Reply by arunnbabu81** [reply]
> the newline characters need removing from the text. you will need the lastest version of tiddlyclip (2.0.01)
> 
> add this line at the begining of pdfFieldModifiers `@text=((*@removelf(@text)*))` and change `$text=|><pdflink.....` to `$text=((*$text*))|><pdflink........`
> 
> you will need this macro
> 
> [removelf.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/8051306/removelf.js.json.zip)

Its not working.


![image](https://user-images.githubusercontent.com/67494083/153672942-40328ce2-6556-4840-86a5-708de2a395e4.png)


![image](https://user-images.githubusercontent.com/67494083/153672574-d616755e-7387-4a4e-a2c0-8667a97936c6.png)

**Reply by buggyj** [reply]
you are missing a `<` on the macro `><pdfink...` you should also remove the indent markup `>`

**Reply by arunnbabu81** [reply]
> you are missing a `<` on the macro `><pdfink...` you should also remove the indent markup `>`

Thank you…now it is working
